x/auth/ante/unordered.go

```diff
@@ -0,0 +1,69 @@
+package ante
+
+import (
+	"crypto/sha256"
+
+	errorsmod "cosmossdk.io/errors"
+	"cosmossdk.io/x/auth/ante/unorderedtx"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+var _ sdk.AnteDecorator = (*UnorderedTxDecorator)(nil)
+
+// UnorderedTxDecorator defines an AnteHandler decorator that is responsible for
+// checking if a transaction is intended to be unordered and if so, evaluates
+// the transaction accordingly. An unordered transaction will bypass having it's
+// nonce incremented, which allows fire-and-forget along with possible parallel
+// transaction processing, without having to deal with nonces.
+//
+// The transaction sender must ensure that unordered=true and a timeout_height
+// is appropriately set. The AnteHandler will check that the transaction is not
+// a duplicate and will evict it from memory when the timeout is reached.
+//
+// The UnorderedTxDecorator should be placed as early as possible in the AnteHandler
+// chain to ensure that during DeliverTx, the transaction is added to the UnorderedTxManager.
+type UnorderedTxDecorator struct {
+	// maxUnOrderedTTL defines the maximum TTL a transaction can define.
+	maxUnOrderedTTL uint64
+	txManager       *unorderedtx.Manager
+}
+
+func NewUnorderedTxDecorator(maxTTL uint64, m *unorderedtx.Manager) *UnorderedTxDecorator {
+	return &UnorderedTxDecorator{
+		maxUnOrderedTTL: maxTTL,
+		txManager:       m,
+	}
+}
+
+func (d *UnorderedTxDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	unorderedTx, ok := tx.(sdk.TxWithUnordered)
+	if !ok || !unorderedTx.GetUnordered() {
+		// If the transaction does not implement unordered capabilities or has the
+		// unordered value as false, we bypass.
+		return next(ctx, tx, simulate)
+	}
+
+	if unorderedTx.GetTimeoutHeight() == 0 {
+		return ctx, errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "unordered transaction must have timeout_height set")
+	}
+
+	if unorderedTx.GetTimeoutHeight() > uint64(ctx.BlockHeight())+d.maxUnOrderedTTL {
+		return ctx, errorsmod.Wrapf(sdkerrors.ErrInvalidRequest, "unordered tx ttl exceeds %d", d.maxUnOrderedTTL)
+	}
+
+	txHash := sha256.Sum256(ctx.TxBytes())
```

Comment:
is there a better way to reuse the hash calculated by the cometbft?

Reply:
Unforunately, we do not have this information exposed on the `types.Context`. I don't even think CometBFT provides it actually.